packages/network-controller/src/create-network-client.ts

```diff
@@ -0,0 +1,188 @@
+import {
+  createAsyncMiddleware,
+  createScaffoldMiddleware,
+  JsonRpcEngine,
+  mergeMiddleware,
+  JsonRpcMiddleware,
+} from 'json-rpc-engine';
+import {
+  createBlockCacheMiddleware,
+  createBlockRefMiddleware,
+  createBlockRefRewriteMiddleware,
+  createBlockTrackerInspectorMiddleware,
+  createInflightCacheMiddleware,
+  createFetchMiddleware,
+  createRetryOnEmptyMiddleware,
+} from '@metamask/eth-json-rpc-middleware';
+import {
+  providerFromEngine,
+  providerFromMiddleware,
+  SafeEventEmitterProvider,
+} from '@metamask/eth-json-rpc-provider';
+import { createInfuraMiddleware } from '@metamask/eth-json-rpc-infura';
+import type { Hex } from '@metamask/utils/dist';
+import { PollingBlockTracker } from 'eth-block-tracker/dist';
+import { NetworksChainId } from '@metamask/controller-utils';
+
+const SECOND = 1000;
+
+export enum NetworkClientType {
+  Custom = 'custom',
+  Infura = 'infura',
+}
+
+type CustomNetworkConfiguration = {
+  chainId?: Hex;
+  rpcUrl: string;
+  type: NetworkClientType.Custom;
+};
+
+// could improve this type... this is not a very complete list
+export type InfuraNetworkType = 'mainnet' | 'goerli' | 'sepolia';
```

Comment:
I don't remember if I made a comment about this before, but we have some network-related stuff in `controller-utils`, specifically the [`NetworkType` enum](https://github.com/MetaMask/core/blob/main/packages/controller-utils/src/types.ts#L4). So instead of redefining this list, perhaps this could be:

``` suggestion
export type InfuraNetworkType = Exclude<NetworkType, 'localhost' | 'rpc'>;
```